### ISSUE TITLE
GridLocations give different results when used as context vs for loop when inside another context

### ISSUE THREAD

**Issue by MatthiasJ1** [gumyr/build123d#391]
According to the ReadTheDocs, these are equivalent uses of Locations yet they give different results. It appears the outer Locations context is applied twice.

```py
with BuildPart() as p:
    Box(50, 10, 1)
    with Locations((-5, 0, 0)):
        with GridLocations(10, 0, 5, 1):
            Box(1, 10, 4)
```
<img width="509" alt="1" src="https://github.com/gumyr/build123d/assets/149267773/2a520600-0060-4b7b-8a74-b6e6691b88ec">

```py
with BuildPart() as p:
    Box(50, 10, 1)
    with Locations((-5, 0, 0)):
        for l in GridLocations(10, 0, 5, 1):
            Box(1, 10, 4, l)
####### or #######
with BuildPart() as p:
    Box(50, 10, 1)
    with Locations((-5, 0, 0)):
        for l in GridLocations(10, 0, 5, 1):
            with Locations(l): Box(1, 10, 4)
```
<img width="509" alt="2" src="https://github.com/gumyr/build123d/assets/149267773/6a960865-5e67-421a-9c6f-7a89c8c47116">

---
This difference only arises when inside another Locations context. If you remove the outer Locations context, all of the below snippets give the same result.

```py
with BuildPart() as p:
    Box(50, 10, 1)
    with GridLocations(10, 0, 5, 1):
        Box(1, 10, 4)
#############################
with BuildPart() as p:
    Box(50, 10, 1)
    for l in GridLocations(10, 0, 5, 1):
        with Locations(l): Box(1, 10, 4)
#############################
with BuildPart() as p:
    Box(50, 10, 1)
    for l in GridLocations(10, 0, 5, 1):
        Box(1, 10, 4, l)
```
<img width="509" alt="3" src="https://github.com/gumyr/build123d/assets/149267773/0662870f-87b8-4fe9-84f9-891ed2feb96a">

**Comment by MatthiasJ1** [comment]
Another example of an issue this creates:

```py
with BuildSketch() as A:
    with Locations((0,0), *GridLocations(5,5,2,2)):
        Circle(1)

with BuildSketch(Pos(10, 0)) as B:
    with Locations((0,0), *GridLocations(5,5,2,2)):
        Circle(1)
```
<img width="396" alt="s" src="https://github.com/gumyr/build123d/assets/149267773/c25b9e3d-2d0f-4927-aefb-7e9e3a259c7f">

**Comment by gumyr** [comment]
Nested Locations are impacted by other Locations within their scope so in your examples (non Sketch - I haven't looked into that yet) you're nesting 3 Locations so the first one is being applied twice.  If you move the GridLocations outside of the scope of the 1st Locations (e.g. outside of any of the `with` blocks) I believe you'll get the results you're looking for.

**Comment by MatthiasJ1** [comment]
Yes, that is correct. I would argue, however, that applying locations contexts non-destructively only upon entering a `with` statement would work better.

The way it currently works, the generated locations are transformed by the active locations contexts upon creation. This means that created location contexts do not accurately represent the parameters they are generated with and are *only* valid in the context in which they are created. This means that there is no consistent hierarchy which leads to buggy behavior like can be seen here.

For example, `GridLocations(2,2,2,2)` claims to create an object which represents 4 corners of a rectangle centered at the current origin. However, if I happen to be inside a `Locations((1,1))`, the `GridLocations` would now represent a rectangle with a lower-left alignment, which is a different `GridLocations` despite having the same constructor arguments. If you remain in the context in which it was created it remains transparently consistent, however it becomes invalid outside that context. You could argue it is intended to only be valid inside the context, however, this can still lead to bugs as is seen in this issue. This is also different from creating a `Rectangle` since the `Rectangle` is the final generated output. You wouldn't create another `Rectangle` inside the context of a `Rectangle`.

My proposed alternative is to have locations contexts represent only the parameters they are specified with, and that the hierarchy of active locations contexts be applied when a when they are "entered" using a `with` block. This would result in the exact same behavior for existing code that follows your recommended methods of using Locations context while also making all of the "currently not working" examples specified in this issue work as expected. It would also be more consistent with, IMO, the intended builder context behavior.

This would mean that a `Locations` context would be valid in its own right and could be reused anywhere else in the code. Say if you had a bolt mounting pattern that you wanted to store in a Locations context. Currently this is impossible using the Builder API since the context stores the entire hierarchy of transformations, not just its own transformations, meaning that it can only represent a single set of locations. My proposal would fix this would breaking existing intended behaviors.

**Comment by gumyr** [comment]
I don't understand what you're proposing here - maybe some code would help.  The current behaviour is as follows:
```py
with BuildPart() as plate:
    Box(120, 50, 10)
    with Locations(plate.faces() >> Axis.Z):
        with GridLocations(70, 0, 2, 1):
            with PolarLocations(10, 4):
                CounterBoreHole(2, 4, 2)
```
![image](https://github.com/gumyr/build123d/assets/85638518/5ab6f939-4772-4e72-9f36-37174b4bf438)

How would your proposal be different in this case?

**Comment by gumyr** [comment]
With respect to the following:
```py
with BuildSketch() as A:
    with Locations((0,0), *GridLocations(5,5,2,2)):
        Circle(1)

with BuildSketch(Pos(10, 0)) as B:
    with Locations((0,0), *GridLocations(5,5,2,2)):
        Circle(1)
```
I don't know of any way for the S/W to know what the use intent is so it's going to have to stay the same.  Here how this should be done:
```py
with BuildSketch() as A:
    with Locations((0,0), *GridLocations(5,5,2,2)):
        Circle(1)

with BuildSketch(Pos(10, 0)) as B:
    with Locations((0,0), *GridLocations(5,5,2,2).local_locations):
        Circle(1)
```
![image](https://github.com/gumyr/build123d/assets/85638518/eb761d9d-2c9c-49a5-baa9-7281bc4a0109)

In this case `local_locations` should be used as they are relative to Workplane and will be adjusted within the Locations call.  `for` loops are definitely 2nd class citizens here as the optimized case is for Locations in `with` blocks.  I've updated the docs to show the `local_locations` attribute.

I've created an Issue (https://github.com/gumyr/build123d/issues/431) to explore if's correct and possible to change what the Location classes return based on if they are in `with` blocks.
